chapters/grid-search.qmd

```diff
@@ -384,7 +383,497 @@ A simple workaround for creating a space-filling design is to repeat the unique
+
+### Parallel Processing {#sec-parallel}
+
+Fortunately, none of the  `r format(nrow(large_sfd) * nrow(sim_rs), big.mark = ",")` models in our large grid depend on one another and can be computed separately. Since almost all modern computers have GPUs and multiple CPUs, we can break the computations into different "chunks" of work and execute them simultaneously on other processors (or separate computers entirely). The parallel processing of models can significantly reduce the time it takes to tune using grid search.
+
+```{r}
+#| label: bst-speeds
+#| include: false
+bst_speedup <- round(sfd_seq_time[3]/sfd_time[3], 1)
+```
+
+For example, consider the space-filling designs with `r nrow(bst_sfd)` candidates evaluated across `r nrow(sim_rs)` resamples. We spread these models  `r nrow(bst_sfd) * nrow(sim_rs)` model fits across `r num_workers` worker processes (on the same computer). Despite the meager computation costs of training each boosted tree, there was still a `r bst_speedup`-fold speedup (`r round(sfd_seq_time[3], 1)`s versus `r round(sfd_time[3], 1)`s) when run in parallel. 
+
+However, there are some nuances related to parallel processing.
+
+First, it is an excellent idea to parallelize the "longest loop" (literally or figuratively). Looking back at @alg-grid-search, there is a loop across the $s$ candidates in line 6. Line 7 contains the resampling loop across $B$ resamples. If the data set is large, it may be optimal to invert the loop where we parallel process across the $B$ resamples and execute the $s$ models within each. Keeping each resample on a single worker means less input/output traffic across the workers. See @tmwr [Section 13.5.2](https://www.tmwr.org/grid-search#parallel-processing). Alternatively, if the data are not large, it could be best to “flatten” the two loops into a single loop with $s\times B$ iterations (with as many workers as possible). 
+
+Additionally, if expensive preprocessing is used, a naive approach where each pipeline is processed in parallel might be counterproductive because we unnecessarily repeat the same preprocessing. For example, suppose we are tuning a supervised model along with UMAP preprocessing. By sending each of the $s$ candidates to a worker, the same UMAP training occurs for each set of candidates that correspond to the supervised model. Instead, a _conditional process_ can be used where we loop across the UMAP tuning parameter combinations and, for each, run another loop across the parameters associated with the supervised model. 
+
+One other consideration is memory. If the data used to train and evaluate the model are copied for each worker, the number of workers can be restricted to fit within system memory. 
+
+In any case, parallel processing can almost always improve the speed of model tuning. 
+
+### Racing {#sec-racing}
+
```

Comment:
I will eventually write summaries of other methods (mostly for deep learning) like hyperopt, spearmint, etc. Not a high priority right now.